evm/src/main/java/org/hyperledger/besu/evm/precompile/BigIntegerModularExponentiationPrecompiledContract.java

```diff
@@ -87,46 +119,68 @@ public PrecompileContractResult computePrecompile(
   }
 
   // Equation to estimate the multiplication complexity.
-  public static BigInteger multiplicationComplexity(final BigInteger x) {
-    if (x.compareTo(BIGINT_64) <= 0) {
+  public static long multiplicationComplexity(final long x) {
+    if (x <= 64) {
       return square(x);
-    } else if (x.compareTo(BIGINT_1024) <= 0) {
-      return square(x).divide(BIGINT_4).add(BIGINT_96.multiply(x)).subtract(BIGINT_3072);
+    } else if (x <= 1024) {
+      return (square(x) / 4) + (x * 96) - 3072;
     } else {
-      return square(x).divide(BIGINT_16).add(BIGINT_480.multiply(x)).subtract(BIGINT_199680);
+      return (square(x) / 16) + (480 * x) - 199680;
     }
   }
 
-  public static BigInteger baseLength(final Bytes input) {
-    return extractParameter(input, BASE_LENGTH_OFFSET, PARAMETER_LENGTH);
+  public static long baseLength(final Bytes input) {
+    return extractParameterLong(input, BASE_LENGTH_OFFSET, PARAMETER_LENGTH);
   }
 
-  public static BigInteger exponentLength(final Bytes input) {
-    return extractParameter(input, EXPONENT_LENGTH_OFFSET, PARAMETER_LENGTH);
+  public static long exponentLength(final Bytes input) {
+    return extractParameterLong(input, EXPONENT_LENGTH_OFFSET, PARAMETER_LENGTH);
   }
 
-  public static BigInteger modulusLength(final Bytes input) {
-    return extractParameter(input, MODULUS_LENGTH_OFFSET, PARAMETER_LENGTH);
+  public static long modulusLength(final Bytes input) {
+    return extractParameterLong(input, MODULUS_LENGTH_OFFSET, PARAMETER_LENGTH);
   }
 
-  private static BigInteger extractParameter(
-      final Bytes input, final int offset, final int length) {
+  public static BigInteger extractParameter(final Bytes input, final int offset, final int length) {
     if (offset > input.size() || length == 0) {
       return BigInteger.ZERO;
     }
     final byte[] raw = Arrays.copyOfRange(input.toArray(), offset, offset + length);
     return new BigInteger(1, raw);
   }
 
-  public static BigInteger extractParameter(
-      final Bytes input, final BigInteger offset, final int length) {
-    if (BigInteger.valueOf(input.size()).compareTo(offset) <= 0) {
-      return BigInteger.ZERO;
+  public static long extractParameterLong(final Bytes input, final int offset, final int length) {
+    if (offset >= input.size() || length == 0) {
+      return 0;
+    }
+    Bytes num;
+    if (offset + length <= input.size()) {
+      num = input.slice(offset, length).trimLeadingZeros();
+    } else {
```

Comment:
Suggestion: the else case behavior wouldn't be intuitive to me as a caller, might warrant a javadoc explaining that you can request more bytes back than available from your offset.